File: 1680-concatenation-of-consecutive-binary-numbers/1680-concatenation-of-consecutive-binary-numbers.cpp
#define MOD 1000000007
class Solution {
public:
    int concatenatedBinary(int n) {
        long long res = 1;
        int mask = 1 << 2;
        char shift = 2;
        for(int i = 2; i <= n; i++)
        {
            if(mask & i)
            {
                shift++;
                mask <<= 1;
            }
            res = (((res << shift) % MOD) + i) % MOD;
        }
        return res;
    }
};

File: 1680-concatenation-of-consecutive-binary-numbers/1680-concatenation-of-consecutive-binary-numbers.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        res = 1
        MOD = 1000000007
        mask = 1 << 2
        shift = 2
        for i in range(2, n + 1):
            if mask & i:
                shift += 1
                mask <<= 1
            res = (((res << shift) % MOD) + i) % MOD
        return res